Week 3/Extra Practice Problems (Week 3)/0087: Scramble String/dominic.cpp

```diff
@@ -0,0 +1,34 @@
+class Solution {
+public:
+    vector<vector<vector<int>>> memo;
+
+    bool solve(string s1, string s2, int l, int r) {  
+        int len = s1.length(); 
+        if (memo[l][r][len] > -1) return memo[l][r][len];
+        if (s1 == s2) return memo[l][r][len] = 1;
+        
+
+        for (int i = 1; i < len; i++) {
+            string x = s1.substr(0, i), y = s1.substr(i, len);
+            string s2x = s2.substr(0, i), s2y = s2.substr(i, len);
+            if (solve(x, s2x, l, r) && solve(y, s2y, l+i, r+i)) {
```

Comment:
This works but it's slow

We can skip a bunch of recursive calls if we make sure both parts have the same counts of letters